tests/js/migration-tests.js

```diff
@@ -0,0 +1,158 @@
+        TestCase.assertEqual(1, count);
+        realm.close();
+
+        // invalid migration function
+        TestCase.assertThrows(function() {
+            new Realm({schema: [], schemaVersion: 2, migration: 'invalid'});
+        });
+
+        // migration function exceptions should propogate
+        var exception = 'expected exception';
+        realm = undefined;
+        TestCase.assertThrowsException(function() {
+            realm = new Realm({schema: [], schemaVersion: 3, migration: function() {
+                throw exception;
+            }});
+        }, exception);
+        TestCase.assertEqual(realm, undefined);
+        TestCase.assertEqual(Realm.schemaVersion(Realm.defaultPath), 1);
+
+        // migration function shouldn't run if nothing changes
+        realm = new Realm({schema: [Schemas.TestObject], migration: migrationFunction, schemaVersion: 1});
+        TestCase.assertEqual(1, count);
+        realm.close();
+
+        // migration function should run if only schemaVersion changes
+        realm = new Realm({schema: [Schemas.TestObject], migration: function() { count++; }, schemaVersion: 2});
+        TestCase.assertEqual(2, count);
+        realm.close();
+    },
+
+    testDataMigration: function() {
+        var realm = new Realm({schema: [{
+            name: 'TestObject',
+            properties: {
+                prop0: 'string',
+                prop1: 'int',
+            }
+        }]});
+        realm.write(function() {
+            realm.create('TestObject', ['stringValue', 1]);
+        });
+        realm.close();
+
+        var realm = new Realm({
+            schema: [{
+                name: 'TestObject',
+                properties: {
+                    renamed: 'string',
+                    prop1: 'int',
+                }
+            }], 
+            schemaVersion: 1, 
+            migration: function(oldRealm, newRealm) {
+                var oldObjects = oldRealm.objects('TestObject');
+                var newObjects = newRealm.objects('TestObject');
+                TestCase.assertEqual(oldObjects.length, 1);
+                TestCase.assertEqual(newObjects.length, 1);
+
+                TestCase.assertEqual(oldObjects[0].prop0, 'stringValue');
+                TestCase.assertEqual(oldObjects[0].prop1, 1);
+                //TestCase.assertThrows(function() { oldObjects[0].renamed; });
+
+                //TestCase.assertThrows(function() { newObjects[0].prop0; });
+                TestCase.assertEqual(newObjects[0].renamed, '');
+                TestCase.assertEqual(newObjects[0].prop1, 1);
+
+                newObjects[0].renamed = oldObjects[0].prop0;
+            }
+        });
+
+        var objects = realm.objects('TestObject');
+        TestCase.assertEqual(objects.length, 1);
+        TestCase.assertEqual(objects[0].renamed, 'stringValue');
+        TestCase.assertEqual(objects[0].prop1, 1);
+        TestCase.assertThrows(function() { newObjects[0].prop0; });
```

Comment:
`objects[0].prop0` is `undefined` in this case which is what we should be testing.